### ISSUE TITLE
Can I put the extensionsGallery on my vscode fork?

### ISSUE THREAD

**Issue by evandrocoan** [microsoft/vscode#31168]
By default the https://github.com/Microsoft/vscode does not come with key:

    "extensionsGallery": {
        "serviceUrl": "",
        "cacheUrl": "",
        "itemUrl": ""
    }

on its https://github.com/Microsoft/vscode/blob/master/product.json file. That entry is required to install extensions from the marketplace as showed on:

1.  https://stackoverflow.com/questions/37143536/no-extensions-found-when-running-visual-studio-code-from-source

Also there are commits, where their developer putted the key.

But sequentially removed it justifying:

> Revert inadvertant change to product.json

Why they are removing it? Seems they are using it while developing, as we may see he let slip the line to the upper stream. Can I put it on my fork? Or I can only use it locally as they seem to be doing?

___

Related issues:

1. https://github.com/Microsoft/vscode/issues/60 Command "Extensions: Install Extension" missing when running vscode from source
1. https://github.com/Microsoft/vscode/issues/2647 Command "Extensions: Install Extension" missing when running vscode from source
1. https://github.com/Microsoft/vscode/issues/28253 Can't see extensions or load via command line
1. https://github.com/Microsoft/vscode/issues/23831 How do I enable extensions in the extensions dir in OSS Dev mode
1. https://github.com/Microsoft/vscode-wiki/pull/30 Add marketplace extension installation details

**Comment by joaomoreno** [comment]
cc @chrisdias @chrmarti

**Comment by MattSturgeon** [comment]
Since b00945f was committed under the MIT license, I would argue that you _**can**_ include any changes from within it in any MIT compatible fork.

While you cannot copy from the `product.json` included in the proprietary VS Code product, I would maintain that the following text has been released under MIT and therefore can be redistributed within your fork.

```json
"extensionsGallery": {
    "serviceUrl": "https://marketplace.visualstudio.com/_apis/public/gallery",
    "cacheUrl": "https://vscode.blob.core.windows.net/gallery/index",
    "itemUrl": "https://marketplace.visualstudio.com/items"
}
```

---

_Being new to the project and conscious of Free Software licenses, I have never downloaded the [propitiatory product](https://code.visualstudio.com/Download), agreed to the [license](https://code.visualstudio.com/License/) nor read any of the propitiatory code. My only knowledge of the above code comes through b00945f and this issue._

_I am not a lawyer and this is not legal advice._

**Comment by kenijo** [comment]
A question on the cacheURL: https://vscode.blob.core.windows.net/gallery/index

It looks like it hasn't been updated since March 2017. The most recent "lastUpdated" date is 2017-03.
It there anyway to get this fixed or to get working URL?

**Comment by dtinth** [comment]
Based on [this comment](https://github.com/Microsoft/vscode/issues/1557#issuecomment-173449135), the answer is yes?

**Comment by chris-kahn** [comment]
Even if it wasn't licensed, the code in question is so tiny and generic (it's literally a few labels and web addresses) that I would question whether it's even copyrightable in the first place.

**Comment by jens1o** [comment]
@chrisdias any legal update?

**Comment by ivanslf** [comment]
Acknowledging some public URLs and copying them into the settings file of your own product isn't a copyright infringement _at all_.

**Comment by nhooyr** [comment]
@ivanslf its not just about copying the lines, the terms of use for the vs code marketplace state that it can only be used with visual studio products. It's not clear whether that includes forks.

**Comment by rrmckinley** [comment]
This exists now
https://projects.eclipse.org/proposals/eclipse-open-vsx-registry
https://devclass.com/2020/03/31/eclipses-theia-sees-1-0-declared-not-your-parents-ide/

> Sven Efftinge, co-founder of both Gitpod and TypeFox... continued, “Microsoft prohibits non-Visual Studio products from installing any binaries downloaded from their marketplace.”
> 
> So, “We encourage VS Code extension developers to push their extensions to Open VSX in addition to Microsoft’s marketplace.” Open VSX is “an open-source implementation of a VS Code extension registry that we have developed under the umbrella of the Eclipse Foundation”.

**Comment by ivanfilhoz** [comment]
@rrmckinley I wasn't aware of that, thanks.

For ethical reasons, I'm uninstalling VSCode immediately in favor of [Theia](https://theia-ide.org/).

> **Vendor Neutral**
> The Theia project is hosted at the Eclipse Foundation, a not-for-profit corporation, and is developed by a diverse community.
> Unlike other "open-source" projects, projects hosted at an Open-Source Foundation are protected against single-vendor decisions against the interest of the diverse community. Learn more [here](https://www.eclipse.org/projects/dev_process/).

**Comment by rrmckinley** [comment]
@ivanfilhoz It was announced today so very few people were aware. 

https://github.com/eclipse-theia/theia/issues/7072
https://github.com/eclipse-theia/theia/commit/73bd544d99a1f93159cc83cedd9769e627033811

If you get the chance, I'd love to hear your first impressions of Theia 1.0, thanks

**Comment by spoenemann** [comment]
> For ethical reasons, I'm uninstalling VSCode immediately in favor of Theia.

@ivanfilhoz Theia is not an editor/IDE, but a _framework_ for building IDEs. There's no generic Theia-based application that you could download at the moment. But of course anyone is free to start a new project using Theia to build an Electron app that mimics VS Code.

**Comment by rolodato** [comment]
The [current ToS for Visual Studio Marketplace](https://cdn.vsassets.io/v/M146_20190123.39/_content/Microsoft-Visual-Studio-Marketplace-Terms-of-Use.pdf) say (emphasis mine):

> [...] Microsoft Visual Studio, Visual Studio for Mac, **Visual Studio Code**, Azure DevOps, Team Foundation Server and successor Microsoft products and services (collectively, the “Visual Studio Products and Services”)
> [...]
> Marketplace Offerings are intended for use only with Visual Studio Products and Services and you may only install and use Marketplace Offerings with Visual Studio Products and Services.

IANAL, not legal advice etc but in my opinion this means that VS Code forks cannot use the Marketplace directly. A fork of VS Code is not VS Code.

**Comment by MattSturgeon** [comment]
> The current ToS for Visual Studio Marketplace say 

Perhaps, but that's only relevant if you've agreed to the ToS.

**Comment by rolodato** [comment]
The terms also say:

> By accessing or using the Marketplace, you represent and warrant to us that you have the authority to accept this Agreement, and you agree to be bound by its terms.

**Comment by debo** [comment]
> The [current ToS for Visual Studio Marketplace](https://cdn.vsassets.io/v/M146_20190123.39/_content/Microsoft-Visual-Studio-Marketplace-Terms-of-Use.pdf) say (emphasis mine):
> 
> > [...] Microsoft Visual Studio, Visual Studio for Mac, **Visual Studio Code**, Azure DevOps, Team Foundation Server and successor Microsoft products and services (collectively, the “Visual Studio Products and Services”)
> > [...]
> > Marketplace Offerings are intended for use only with Visual Studio Products and Services and you may only install and use Marketplace Offerings with Visual Studio Products and Services.
> 
> IANAL, not legal advice etc but in my opinion this means that VS Code forks cannot use the Marketplace directly. A fork of VS Code is not VS Code.

Why fork it then? Wouldn't it be best to just clone vscode, add the official marketplace and compile that? Assuming you can because I still didn't manage to be fair. It would be nice to have an "official" straight forward and clear answer from Microsoft on this.

**Comment by debo** [comment]
@gasinvein that's my thought exactly, hence my comment on the quoted one. 

However, [this message](https://github.com/Microsoft/vscode/issues/1557#issuecomment-173449135) from one of the core contributors seems to state otherwise so I'm a bit confused to be fair.

**Comment by spoenemann** [comment]
This article states very clearly that "Visual Studio Code" and "Code - OSS" are different things:
https://github.com/microsoft/vscode/wiki/Differences-between-the-repository-and-Visual-Studio-Code

See also #60.

**Comment by debo** [comment]
Absolutely, but when why a core developer, Microsoft representative, would say otherwise, that you can use the marketplace with Code - OSS?

**Comment by soredake** [comment]
Still no answer for this?

**Comment by taylorjdawson** [comment]
What is the official verdict?

**Comment by ankit2web** [comment]
I think I need to uninstall `vscode` effective immediately considering this issue is open since 2017 & the legality of this issue hasn't been yet decided.

**Comment by therealdandecker** [comment]
I interpret from this link [this link ](https://github.com/microsoft/vscode/wiki/Differences-between-the-repository-and-Visual-Studio-Code) shared earlier in this thread that the Visual Studio Marketplace is closed source and its not legal to use the marketplace in another distribution of the open source code.  Microsoft explicitly states it in their vscode repo.

**Comment by tajetaje** [comment]
Any update on any internal discussion on this in the last fourteen months? Even a simple "we can't give any legal advice here" would be fine, but unless I missed someone nobody from Microsoft has said anything for almost FIVE YEARS. Any information at all would be much appreciated!

**Comment by toddewilliams** [comment]
Since Microsoft won't directly interpret their Visual Studio Marketplace Terms of Use for you, I will since I believe they're quite clear:

_Short Answer:_ **NO, YOU CANNOT LEGALLY USE THE MICROSOFT MARKETPLACE IN ANY NON-MICROSOFT PRODUCT.**

_Long Answer:_ To see this for yourself, begin by selecting the [Terms of Use](https://cdn.vsassets.io/v/M190_20210811.1/_content/Microsoft-Visual-Studio-Marketplace-Terms-of-Use.pdf) link at the bottom of the [Marketplace](https://marketplace.visualstudio.com) page.

In Section 1b of the Terms of Use, please note the following prohibition stating how the marketplace may legally be used:
**`Marketplace Offerings are intended for use only with In-Scope Products and Services and you may install and use Marketplace Offerings only with In-Scope Products and Services.`**

OK, so you can only use the marketplace to install software for use by "in-scope products and services".  So what are those?

That is answered in the introductory paragraph of the Terms of Use:
`The Marketplace enables you to access or purchase products or services which are designed to work with and extend the capabilities of Microsoft Visual Studio, Visual Studio for Mac, Visual Studio Code, GitHub Codespaces, Azure DevOps, Azure DevOps Server, and successor products and services (the “In-Scope Products and Services”) offered by us and GitHub, Inc. (“GitHub”).`

The services are explicitly listed, and must come from us (Microsoft) and GitHub. 

In addition to this, the vast majority of the Microsoft extensions have additional license agreements that also explicitly forbid their use in non-Microsoft products. For example, the [license for Live Share](https://marketplace.visualstudio.com/items/MS-vsliveshare.vsliveshare/license) is quite clear in Section 1:
`INSTALLATION AND USE RIGHTS. You may install and use any number of copies of the software to use solely with Microsoft Visual Studio family of products.`

_So - "Can I put the extensionsGallery on my vscode fork?"_
No, you absolutely may not and you may not use the marketplace to install software into your fork. The entire reason [OpenVSX](https://open-vsx.org) exists is because using Microsoft's marketplace for non-Microsoft products is expressly illegal.

Perhaps someone with access can now mark this issue as 'Resolved'.

**Comment by theAkito** [comment]
@toddewilliams 

Thanks for the effort. This clarified everything and I think we all can understand now, without a doubt. 😃

**Comment by toddewilliams** [comment]
Hey @chrisdias, I believe you can close this now,  unless you want to officially contradict the [analysis in my prior comment](https://github.com/microsoft/vscode/issues/31168#issuecomment-1136375670).

**Comment by StarNumber12046** [comment]
as far as I know, if you HAVE vscode, you can use the marketplace BUT not distribute it

**Comment by sbussard** [comment]
Way to Microsoft, Microsoft

**Comment by GitMensch** [comment]
Hm, this is still open, not resolved...
Actually @chrisdias wrote in https://github.com/Microsoft/vscode/issues/1557#issuecomment-173449135

> yes, you can configure product.json to point to the official marketplace. look in the product.json that is installed with the visual studio code distro for the url.

So that _does_ contradict to @toddewilliams analysis.
As others pointed out: it is totally fine to have extensions published in the MS marketplace that are explicit licensed only for use with Microsoft Visual Code - but what would be the point of the marketplace to forbid its use with OSS vscode, especially as it is also allowed to be accessed via any browser?

As vscode is the only OSS the list could easily be adjusted to read;

```diff
-Visual Studio Code, GitHub Codespaces, Azure DevOps, Azure DevOps Server, and successor products and services
+Visual Studio Code and derivatives thereof, GitHub Codespaces, Azure DevOps, Azure DevOps Server, and successor products and services
```

... it may be good to update the marketplace TOU as the use is said to be only valid 

>  for use only with In-Scope Products and Services

where does accessing it with the browser instead of the APIs fit there?

**Comment by toddewilliams** [comment]
@chrisdias is not in a position to speak for Microsoft and was likely told he overstepped his role in his prior comment. That is why he has not commented further. 

And while it is true that you **can** configure project.json to point to the official marketplace that doesn't mean you **may** (ie. you legally are permitted to do it). Licenses tell you what you are legally permitted to do and they are very clear on this point; I stand by my [prior analysis, above](https://github.com/microsoft/vscode/issues/31168#issuecomment-1136375670).

Microsoft benefits from not answering this question directly as the only answer is to admit that you are not permitted to use the their official marketplace in anything that isn't a Microsoft product. They don't want to do that for obvious reasons.

I continue to invite an actual representative of Microsoft to contradict my prior analysis and close this issue if I've made any errors in the interpretation of their licenses.

**Comment by GitMensch** [comment]
> Marketplace terms of use https://cdn.vsassets.io/v/M146_20190123.39/_content/Microsoft-Visual-Studio-Marketplace-Terms-of-Use.pdf state that "Marketplace 
> Offerings are intended for use only with Visual Studio Products and Services and you may only install 
> and use Marketplace Offerings with Visual Studio Products and Services. "
> 
> Since VSCodium is not a Visual Studio product it is not allowed to use VS Marketplace from VSCodium. We will also answer that question soon. 

 _Originally posted by @isidorn in [#231569](https://github.com/microsoft/vscode/issues/231569#issuecomment-2531211937)_

@isidorn What does this "soon" refer to?

It's obviously not something like "I'll run some errands and bring you what you need, soon" - I hope it's not something like "We'll soon be walking on the moon again (or even mars)".

**Comment by chrisdias** [comment]
Thanks for bringing this topic up again. Since I am the originator of the confusing comment 10 years ago, I thought it only right that I try to make a definitive statement on this. 

So just a little bit of history. In 2015 (again, 10 years ago!) we were all much younger and less experienced, running fast and figuring out how to do development in the open. At the time, there were no mainstream forks of VS Code. Forks were used by contributors to make the product better, fixing issues, making feature requests, or figuring out how extensions and the API worked. The community was actively contributing, helping to make a good product great (thank you!).

My response was intended to unblock those people who were contributing back to the product. Let’s say they were trying to contribute a fix to the extension viewlet or suggest a new feature for managing extensions. Having a real Marketplace endpoint to test with makes sense.

My response was not intended to say you could circumvent the Marketplace license terms for new products built on a fork of the repository. I’m not a lawyer. I am/was a developer trying to help other developers. In hindsight, and now with much more experience (and many scars :)), I wasn’t very clear on this, and I apologize for the confusion.

So, let’s (hopefully!) clear things up once and for all. Yes, Todd does a good job explaining the Marketplace license and as he asserts, no you cannot use the Marketplace from a fork of VS Code or any product that is not in the VS Family of products. We will however explore carving out a provision in the Marketplace terms to allow you to do this _for testing purposes only_, my original intent. As you can imagine it takes time to update licenses so I don’t know if or when this will happen, but I've already started the discussion. 

If you want more details on the “why”, we recently updated our FAQ (see https://code.visualstudio.com/docs/supporting/faq#_i-cant-access-the-extension-marketplace-with-product-x).

Of course, it’s much more spicey to think we are up to no good and say that we are obligated to staff and run a full-scale global service for everyone to use. Beyond the fact that it doesn’t make much sense business wise, the real but boring reasons we don’t do this are for the safety of our users and the stability of our products.

We build and run not just a service for hosting and downloading extensions, but we employ numerous engineers and product managers to manage the security of publishers, extensions and the users of the VS family of products. We have no control over forks. We don’t know how they access the site, we can’t quickly make changes to our APIs without going through some sort of committee process for our own service, we don’t know what forks do with internal APIs (maybe they break extensions!), we don’t know if they have the latest security fixes, we don’t know what might be in the middle between the fork and the Marketplace (we sign and validate all downloads, see https://code.visualstudio.com/updates/v1_75#_vs-marketplace-extension-signing), we have no control over malicious extensions in these forks, etc. etc. I could go on.

The bottom line is that security, stability, and agility are never spicey until they bite you, which is why they are our top concerns. I hope this helps.
